### ISSUE TITLE
Integrated Recovery Function

### ISSUE THREAD

**Issue by Leurak** [Leurak/MEMZ#17]
I should add a way to restore the system, when it boots into Nyan Cat without the use of external media.

I guess I'll implement some hidden shortcut or something to activate this.

**Comment by Thesola10** [comment]
I think the Windows recovery partitions have a backup of the MBR so if we get pointers for it that could help

**Comment by Leurak** [comment]
It overwrites more than just the MBR. Also there isn't a recovery partition on XP.

My idea is to do the backup onto some non-used area of the hard drive, which could be on some unallocated space or at the last sector of the disk (if that's even accessable by the BIOS Interrupts).

I don't know if the hidden areas before the first partition are already used by the bootloader, I hope it's not so.

Implementing file systems would be way too hard, especially non-standard ones like NTFS, at least for the MBR payload. UEFI has file APIs I think.

**Comment by Thesola10** [comment]
Maybe write the backup while in the OS, save the pointers and patch them into the payload just before writing it to MBR. Pointer saving is what Windows does with hiberfil.sys

**Comment by Thesola10** [comment]
wait... do you happen to have Cygwin by any chance? We're gonna need a linker for leaving free space in the MBR and patching it, like a `.addr` section

**Comment by Thesola10** [comment]
I'm gonna fork&branch to draft a linux-friendly linked system

**Comment by EvanCase54** [comment]
I think there was this one viewer made malware video Dan did that had this kinda feature. Maybe find that and figure out how it's done?

**Comment by Leurak** [comment]
@EvanCase54 Yeah, I know. I guess I could just put it into some unused area of the HDD.

**Comment by moonheart08** [comment]
A file pointer would be the way to go @Leurak 
Plus telling the patcher what to patch isn't that hard. just have a symbol in the MBR's data that points to the patch location (this would require stage2 to either (A) support a format that can store symbols or (B) have some sort of MEMZ specific format for the patch location)

**Comment by Leurak** [comment]
I know. I just didn't check yet if you can force the file to not get fragmented and how you would get a raw pointer to the sector. Best way would probably just using an unused area inside the partition padding.

**Comment by moonheart08** [comment]
You can prevent fragmentation simply by writing the entire file at once instead of modifying it over time. The file growing larger is what causes fragmentation, not it's creation.

**Comment by Leurak** [comment]
But if the disk is aleady fragmented, the file could get split. There might be a minimum fragment size though.
